src/test/java/org/folio/rest/api/ItemEffectiveLocationTest.java

```diff
@@ -0,0 +1,352 @@
+    UUID holdingsRecordId = createInstanceAndHolding(mainLibraryLocationId, annexLibraryLocationId);
+
+    CompletableFuture<HttpClientResponse> createCompleted = new CompletableFuture<>();
+    Item item = buildItem(holdingsRecordId, null, null);
+
+    client
+      .post(InterfaceUrls.itemsStorageUrl(""), item,
+        StorageTestSuite.TENANT_ID, createCompleted::complete);
+
+    HttpClientResponse response = createCompleted.get(5, TimeUnit.SECONDS);
+
+    assertThat(response.statusCode(), is(201));
+    assertThat(response.getHeader("location"), not(isEmptyString()));
+  }
+
+  @Test
+  public void canCalculateEffectiveLocationWhenItemAssociatedToAnotherHolding() throws Exception {
+    UUID initialHoldingsRecordId = createInstanceAndHolding(mainLibraryLocationId, annexLibraryLocationId);
+    UUID updatedHoldingRecordId = createInstanceAndHolding(onlineLocationId, secondFloorLocationId);
+
+    Item item = buildItem(initialHoldingsRecordId, null, null);
+    createItem(item);
+
+    Item itemFetched = getItem(item.getId());
+    assertEquals(itemFetched.getEffectiveLocationId(), annexLibraryLocationId.toString());
+
+    itemsClient.replace(UUID.fromString(itemFetched.getId()),
+      JsonObject.mapFrom(itemFetched).copy()
+        .put("holdingsRecordId", updatedHoldingRecordId.toString())
+    );
+
+    assertEquals(getItem(item.getId()).getEffectiveLocationId(), secondFloorLocationId.toString());
+  }
+
+  @Test
+  public void canCalculateEffectiveLocationWhenItemHasPermLocationAndAssociatedToAnotherHolding() throws Exception {
+    UUID initialHoldingsRecordId = createInstanceAndHolding(mainLibraryLocationId, annexLibraryLocationId);
+    UUID updatedHoldingRecordId = createInstanceAndHolding(secondFloorLocationId);
+
+    Item item = buildItem(initialHoldingsRecordId, onlineLocationId, null);
+    createItem(item);
+
+    Item itemFetched = getItem(item.getId());
+    assertEquals(itemFetched.getEffectiveLocationId(), onlineLocationId.toString());
+
+    itemsClient.replace(UUID.fromString(itemFetched.getId()),
+      JsonObject.mapFrom(itemFetched).copy()
+        .put("holdingsRecordId", updatedHoldingRecordId.toString())
+    );
+
+    assertEquals(getItem(item.getId()).getEffectiveLocationId(), onlineLocationId.toString());
+  }
+
+  private Item getItem(String id) throws Exception {
+    return itemsClient.getById(UUID.fromString(id)).getJson().mapTo(Item.class);
+  }
+
+  /** Store a permanent location UUID and a temporary location UUID. */
+  private static class PermTemp {
+    /** permanent location UUID */
+    UUID perm;
+    /** temporary location UUID */
+    UUID temp;
+    /**
+     * @param perm permanent location UUID
+     * @param temp temporary location UUID
+     */
+    PermTemp(UUID perm, UUID temp) {
+      this.perm = perm;
+      this.temp = temp;
+    }
+    /**
+     * The first character of the perm UUID and of the temp UUID; use _ for null.
```

Comment:
The code has been changed to use locationIdToNameMap. Please update the first javadoc sentence accordingly.